contracts/transfer/src/wasm/callable.rs

```diff
@@ -0,0 +1,356 @@
+        internal::extend_pi_jubjub_affine(&mut pi, &crossover_commitment);
+        internal::extend_pi_jubjub_affine(&mut pi, &crossover_pk);
+        internal::extend_pi_jubjub_affine(
+            &mut pi,
+            &message.value_commitment().into(),
+        );
+        internal::extend_pi_jubjub_scalar(&mut pi, &message.nonce());
+        internal::extend_pi_jubjub_affine(&mut pi, &pk.as_ref().into());
+        message
+            .cipher()
+            .iter()
+            .for_each(|c| internal::extend_pi_bls_scalar(&mut pi, c));
+
+        //  1. S_a↦.append((pk, R))
+        //  2. M_a↦.M_pk↦.append(M)
+        if self.push_message(address, pk, r, message).is_err() {
+            return false;
+        }
+
+        //  3. if a.isPayable() → true, obf, psk_a? then continue
+        //  TODO
+
+        //  4. verify(C.c, M, pk, π)
+        //  TODO
+        let vk = keys::stco();
+        let (_, _, _, _) = (pi, label, spend_proof, vk);
+
+        //  5. C←(0,0,0)
+        //  TODO
+
+        true
+    }
+
+    fn withdraw_from_obfuscated(
+        &mut self,
+        address: BlsScalar,
+        message: Message,
+        r: JubJubAffine,
+        pk: PublicKey,
+        note: Note, // FIXME nothing is done with this note
+        input_value_commitment: JubJubAffine,
+        spend_proof: Vec<u8>,
+    ) -> bool {
+        let scalars =
+            (1 + BlsScalar::SIZE) * (1 + PoseidonCipher::cipher_size());
+        let points = (1 + JubJubAffine::SIZE) * 4;
+
+        let mut pi = Vec::with_capacity(scalars + points);
+        let label = "transfer-withdraw-from-obfuscated";
+        internal::extend_pi_jubjub_affine(&mut pi, &input_value_commitment);
+        internal::extend_pi_jubjub_affine(
+            &mut pi,
+            &message.value_commitment().into(),
+        );
+        internal::extend_pi_jubjub_scalar(&mut pi, &message.nonce());
+        internal::extend_pi_jubjub_affine(&mut pi, &pk.as_ref().into());
+        message
+            .cipher()
+            .iter()
+            .for_each(|c| internal::extend_pi_bls_scalar(&mut pi, c));
+        internal::extend_pi_jubjub_affine(
+            &mut pi,
+            &note.value_commitment().into(),
+        );
+
+        //  1. a ∈ M↦
+        //  2. pk ∈ M_a↦
+        //  3. M_a↦.delete(pk)
+        // FIXME This message is taken and nothing is verified with it
+        let _message = match self.take_message_from_address_key(&address, &pk) {
+            Ok(m) => m,
+            Err(_) => return false,
+        };
+
+        //  4. if |M_c|=1 then S_a↦.append((pk_c, R_c))
+        //  5. if |M_c|=1 then M_a↦.M_pk↦.append(M_c)
+        if self.push_message(address, pk, r, message).is_err() {
+            return false;
+        }
+
+        //  6. if a.isPayable() → true, obf, psk_a? then continue
+        //  TODO
+
+        //  7. verify(c, M_c, No.c, π)
+        //  TODO
+        let vk = keys::wdfo();
+        let (_, _, _, _) = (pi, label, spend_proof, vk);
```

Comment:
Ditto

Reply:
The pending part here is the proof verification

Reply:
https://github.com/dusk-network/rusk/issues/194